File: Simulation/Car.cpp
#include <cmath>
#include <algorithm>
#include <iostream>
#include <map>
#include <vector>
#include <random>
#include <array>

#include <box2d/box2d.h>
#include <SDL_gpu.h>
#include <SDL_FontCache.h>

#include "car.hpp"
#include "..\Draw.hpp"

class RayCastCallback : public b2RayCastCallback
{
public:
    RayCastCallback(b2Filter* filter)
    {
        this->filter = filter;
    }

    float ReportFixture(b2Fixture* fixture, const b2Vec2& point, const b2Vec2& normal, float fraction) override
    {
        const auto filter{ fixture->GetFilterData() };
        if ((this->filter->maskBits & filter.categoryBits) != 0 and (filter.maskBits & this->filter->categoryBits) != 0)
        {
            this->valid = true;
            this->point = point;
            return fraction;
        }

        return -1;
    }

    b2Filter* filter{ nullptr };
    bool valid{ false };
    b2Vec2 point{ 0.0, 0.0 };
};

Car::Car(b2World* world, b2Body* ground, const b2Vec2& position)
{
    //this->sensors.clear();

    //this->sensors.emplace(180, 0.0f);

   //for (auto n{ -8 }; n <= 8; ++n)
   //{
   //    this->sensors.emplace(180.0f / 16.0f * n, 0.0f);
   //}

   //for (auto n{ -2 }; n <= 2; ++n)
   //{
   //    this->sensors.emplace(180.0f / 4.0f * n, 0.0f);
   //}

    this->world = world;
    this->ground = ground;
    this->createBody(position);
    this->step();
}

Car::Car(const Car& other) : Car{ other.world, other.ground, other.body->GetPosition() }
{
}

Car::~Car()
{
    this->world->DestroyBody(this->body);
}

auto Car::createBody(const b2Vec2& position) -> void
{
    { // Body
        b2BodyDef bd{};
        bd.type = b2_dynamicBody;
        bd.position = position;
        bd.angle = b2_pi;
        bd.linearDamping = 0.1f;
        bd.angularDamping = 0.1f;
        bd.userData = const_cast<char*>("car");

        this->body = this->world->CreateBody(&bd);
    }

    { // Chassis
        b2PolygonShape polygon{};
        polygon.SetAsBox(0.075f, 0.128f);

        b2FixtureDef fd{};
        fd.shape = &polygon;
        fd.filter.categoryBits = 0x0002;
        fd.filter.maskBits = 0x0001;
        fd.density = 10.0f;
        fd.friction = 0.1f;
        fd.userData = const_cast<char*>("chassis");

        this->body->CreateFixture(&fd);
    }

    //{ // Direction symbol (triangle)
    //    b2PolygonShape triangle{};
    //    const b2Vec2 vertices[3]{
    //        b2Vec2{ -0.03f, -0.03f },
    //        b2Vec2{ +0.03f, -0.03f },
    //        b2Vec2{ 0.0f, +0.03f },
    //    };
    //    triangle.Set(vertices, 3);
    //
    //    b2FixtureDef fd{};
    //    fd.shape = &triangle;
    //    fd.isSensor = true;
    //    fd.filter.categoryBits = 0x0002;
    //    fd.filter.maskBits = 0x0001;
    //    fd.userData = const_cast<char*>("chassis");
    //
    //    this->body->CreateFixture(&fd);
    //}

    { // Sensors
        for (const auto& [angle, position, distance] : this->sensors)
        {
            b2PolygonShape polygon{};
            polygon.SetAsBox(0.0125f, 0.0125f, position, 0);
    
            b2FixtureDef fd{};
            fd.shape = &polygon;
            fd.isSensor = true;
            fd.filter.categoryBits = 0x0002;
            fd.filter.maskBits = 0x0001;
            fd.userData = const_cast<char*>("sensor");
    
            this->body->CreateFixture(&fd);
        }
    }

    { // Top-down friction
        const auto gravity{ 10.0f };
        const auto inertia{ this->body->GetInertia() };
        const auto mass{ this->body->GetMass() };
        const auto radius{ b2Sqrt(2.0f * inertia / mass) };

        b2FrictionJointDef jd{};
        jd.bodyA = this->ground;
        jd.bodyB = this->body;
        jd.localAnchorA = b2Vec2{ 0.0f, 0.0f };
        jd.localAnchorB = this->body->GetLocalCenter();
        jd.collideConnected = true;
        jd.maxForce = 0.1f * mass * gravity;
        jd.maxTorque = 0.1f * mass * radius * gravity;
        jd.userData = const_cast<char*>("friction");

        this->world->CreateJoint(&jd);
    }
}

auto Car::step() -> void
{
    /*
    for (auto&& [angle, distance] : this->sensors)
    {
        const auto radians{ (angle / 180.0f) * b2_pi };
        stepSensor(&distance, radians);
    }
    */

    stepBody();

 }

auto Car::render(GPU_Target* target) -> void
{
    GPU_SetLineThickness(0.01f);

    renderBody(target);
    for (auto&& [angle, position, distance] : sensors)
    {
        const auto radians{ (angle / 180.0f) * b2_pi };
        renderSensor(target, position, &distance, radians);
    }
}

auto Car::position() const->b2Vec2
{
    return this->body->GetPosition();
}

auto Car::angle() const->float
{
    return this->body->GetAngle();
}

auto Car::doMove(Move move) -> void
{
    if (this->ready)
    {
        this->move = move;
    }
}

auto Car::isReady() const -> bool 
{
    return this->ready;
}

auto Car::isStuck() const -> bool
{
    return this->stuck;
}

auto Car::distances() const -> std::array<std::pair<int, float>, 18>
{
    auto distances{ std::array<std::pair<int, float>, 18>{} };
    for (auto n{ 0 }; n < distances.size(); ++n)
    {
        distances[n].first = std::get<0>(this->sensors[n]);
        distances[n].second = std::get<2>(this->sensors[n]);
    }
    return distances;
}

auto Car::color() const -> Color
{
    return Color::BLACK;
}

auto Car::giroscope() const->std::vector<float>
{
    return {};
}

auto Car::acelerometer() const->std::vector<float>
{
    return {};
}

auto Car::stepBody() -> void
{
    this->body->SetAngularVelocity(0.0f);
    this->body->SetLinearVelocity({ 0.0f,0.0f });

    if (this->move == Move::ROTATE_LEFT)
    {
        this->body->SetAngularVelocity(-4.0f);
    }
    else if (this->move == Move::ROTATE_RIGHT)
    {
        this->body->SetAngularVelocity(+4.0f);
    }
    else if (this->move == Move::MOVE_FORWARD)
    {
        const auto point{ this->body->GetWorldVector(b2Vec2{ 0.0f, +1.0f }) };
        this->body->SetLinearVelocity(point);
    }
    else if (this->move == Move::MOVE_BACKWARD)
    {
        const auto point{ this->body->GetWorldVector(b2Vec2{ 0.0f, -1.0f }) };
        this->body->SetLinearVelocity(point);
    }

    this->move = Move::STOP;
}

auto Car::stepSensor(const b2Vec2& position, float* distance, float angle) -> void
{
    b2Filter filter{};
    filter.categoryBits = 0x0002;
    filter.maskBits = 0x0001;

    const auto maxDistance{ 2.0f };
    const auto start{ this->body->GetWorldPoint(position) };
    const auto end{ this->body->GetWorldPoint(position + b2Mul(b2Rot{ -angle }, b2Vec2{ 0.0f, maxDistance })) };

    auto rd{ std::random_device{} };
    auto mt{ std::mt19937{rd()} };

    auto callback{ RayCastCallback{&filter} };
    this->world->RayCast(&callback, start, end);
    if (callback.valid)
    {
        auto dist{ std::uniform_real_distribution{ -0.03f, +0.03f } };
        const auto noise{ dist(mt) };

        const auto measuredDistance{ b2Distance(start, callback.point) };
        *distance = std::clamp(measuredDistance + noise, 0.0f, 2.0f);
    }
}

auto Car::renderBody(GPU_Target* target) const -> void
{
    Draw::draw(target, this->body);
}

auto Car::renderSensor(GPU_Target* target, const b2Vec2& position, float* distance, float radians) -> void
{
    this->stepSensor(position, distance, radians);

    const auto start{ this->body->GetWorldPoint(position) };
    const auto end{ this->body->GetWorldPoint(position + b2Mul(b2Rot{ -radians }, b2Vec2{ 0.0f, *distance })) };

    // Crosshair
    //GPU_Line(target, end.x - 0.05f, end.y, end.x + 0.05f, end.y, sensorColor);
    //GPU_Line(target, end.x, end.y - 0.05f, end.x, end.y + 0.05f, sensorColor);

    // Line
    //GPU_Line(target, start.x, start.y, end.x, end.y, sensorColor);
}


auto Car::linearVelocity() const -> float
{
    return this->body->GetLinearVelocity().Length();
}

auto Car::angularVelocity() const -> float
{
    return this->body->GetAngularVelocity();
}

File: scripts/car.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

from math import *
from typing import *
from enum import Enum
from Box2D import b2

class Move(Enum):
    STOP = 0
    MOVE_FORWARD = 1
    MOVE_BACKWARD = 2
    ROTATE_LEFT = 3
    ROTATE_RIGHT = 4

class Color(Enum):
    BLACK = 0
    RED = 1
    GREEN = 2
    BLUE = 3
    YELLOW = 4
    MAGENTA = 5
    CYAN = 6
    WHITE = 7

class SensorCallback(b2.rayCastCallback):
    filter: b2.filter = None
    valid: bool = False
    point: b2.vec2 = b2.vec2(0,0)
    
    def __init__(self, filter: b2.filter):
        super(SensorCallback, self).__init__()
        self.filter = filter
        self.valid = False
        self.point = b2.vec2(0,0)
        return

    def ReportFixture(self, fixture: b2.fixture, point: b2.vec2, normal: b2.vec2, fraction: float):
        filter = fixture.filterData
        if ((self.filter.maskBits & filter.categoryBits) != 0 and (filter.maskBits & self.filter.categoryBits) != 0):
            self.valid = True
            self.point = b2.vec2(point)
            return fraction
        return -1

class Car(object):
    __world: b2.world = None
    __ground: b2.body = None
    __body: b2.body = None
    __move = Move.STOP
    __collision = False
    __sensors = {
        0: 0.0
    }
    __sensorsBodies = {
        0: None
    }

    def __init__(self, world: b2.world, ground: b2.body, position: b2.vec2):
        self.__world = world
        self.__ground = ground
        self.__createBody(position)
        return

    def __createBody(self, position: b2.vec2) -> None:
        if (self.__body != None):
            self.__world.DestroyBody(self.__body)

        self.__body = self.__world.CreateBody(
            type = b2.dynamicBody,
            position = position,
            angle = b2.pi,
            angularDamping = 6.0,
            userData = 'car'
        )

        WIDTH = 0.2
        HEIGHT = 0.2

        self.__body.CreateFixture(
            shape = b2.polygonShape(
                vertices=[
                    (0, 0),
                    (WIDTH, 0),
                    (WIDTH, HEIGHT),
                    (0, HEIGHT)
                ]
            ),
            density = 2.0,
            restitution = 0.5,
            categoryBits = 0x0002,
            maskBits = 0x0001,
            userData = (
                'chassis',
                (64, 64, 255, 255),
                (64, 64, 255, 127)
            )
        )

        #self.__body.CreateFixture(
        #    shape = b2.polygonShape(
        #        vertices=[
        #            (WIDTH / 2, HEIGHT / 3),
        #            (2 * WIDTH / 3, 2 * HEIGHT / 3),
        #            (1 * WIDTH / 3, 2 * HEIGHT / 3)
        #        ]
        #    ),
        #    isSensor=True,
        #    categoryBits = 0x0002,
        #    maskBits = 0x0001,
        #    userData = (
        #        'direction',
        #        (0, 0, 255, 255),
        #        (0, 0, 255, 127)
        #    )
        #)

        GRAVITY = 10.0
        INERTIA = self.__body.inertia
        MASS = self.__body.mass
        RADIUS = sqrt(2.0 * INERTIA / MASS)

        self.__world.CreateFrictionJoint(
            bodyA=self.__ground,
            bodyB=self.__body,
            localAnchorA=(0, 0),
            localAnchorB=self.__body.localCenter,
            collideConnected=True,
            maxForce = 1.35 * MASS * GRAVITY,
            maxTorque = 8.975 * MASS * RADIUS * GRAVITY,
            userData='friction'
        )

        for angle in self.__sensorsBodies:
            self.__sensorsBodies[angle] = self.__world.CreateBody(
                type = b2.staticBody,
                position = (0,0),
                userData = 'sensor'
            )
            self.__sensorsBodies[angle].CreateFixture(
                shape = b2.edgeShape(
                    vertices=[
                        (-0.05, 0),
                        (+0.05, 0)
                    ]
                ),
                isSensor = True,
                userData = (
                    'sensor',
                    (0, 255, 0, 255),
                    (0, 255, 0, 64)
                )
            )
            self.__sensorsBodies[angle].CreateFixture(
                shape = b2.edgeShape(
                    vertices=[
                        (0, -0.05),
                        (0, +0.05)
                    ]
                ),
                isSensor = True,
                userData = (
                    'sensor',
                    (0, 255, 0, 255),
                    (0, 255, 0, 64)
                )
            )

        return

    def reset(self) -> None:
        self.__body.SetLinearVelocity(( 0, 0 ))
        self.__body.SetAngularVelocity(0)
        self.__body.SetTransform(( 0, 0 ), 0)

        self.__move = Move.STOP
        self.__collision = False
    
        return

    def step(self) -> None:
        self.__body.ApplyTorque(1.85, True)

        for angle in self.__sensors:
            radians = (angle / 180.0) * b2.pi
            
            maxDistance = 4.0
            start = self.__body.GetWorldPoint(b2.mul(b2.rot( radians ), b2.vec2( 0.0, 0.10 )))
            end = self.__body.GetWorldPoint(b2.mul(b2.rot( radians ), b2.vec2( 0.0, 0.10 + maxDistance )))

            filter = b2.filter(
                categoryBits = 0x0002,
                maskBits = 0x0001
            )
            callback = SensorCallback(filter)
            callback.point = start

            self.__world.RayCast(callback, start, end)
            if (callback.valid):
                self.__sensorsBodies[angle].position = callback.point
            else:
                distance = 9999.9
                self.__sensorsBodies[angle].position = (9999.9,9999.9)

        return

    